kem/mceliece/templates/mceliece.templ.go

```
@@ -0,0 +1,896 @@
+	unusedBits            = 16 - gfBits
+	sysN                  = {{.Param.SysN}}
+	condBytes             = (1 << (gfBits - 4)) * (2*gfBits - 1)
+	irrBytes              = sysT * 2
+	pkNRows               = sysT * gfBits
+	pkNCols               = sysN - pkNRows
+	pkRowBytes            = (pkNCols + 7) / 8
+	syndBytes             = (pkNRows + 7) / 8
+	PublicKeySize         = {{.Param.PublicKeySize}}
+	PrivateKeySize        = {{.Param.PrivateKeySize}}
+	CiphertextSize        = {{.Param.CiphertextSize}}
+	SharedKeySize         = 32
+	seedSize              = 32
+	encapsulationSeedSize = 48
+)
+
+type PublicKey struct {
+	pk [PublicKeySize]byte
+}
+
+type PrivateKey struct {
+	sk [PrivateKeySize]byte
+}
+
+type (
+	gf       = {{.Param.Gf}}.Elt
+	randFunc = func(pool []byte) error
+)
+
+// KEM Keypair generation.
+//
+// The structure of the secret key is given by the following segments:
+// (32 bytes seed, 8 bytes pivots, IRR_BYTES bytes, COND_BYTES bytes, SYS_N/8 bytes).
+// The structure of the public key is simple: a matrix of PK_NROWS times PK_ROW_BYTES bytes.
+//
+// `entropy` corresponds to the l-bit input seed in SeededKeyGen from the specification.
+// The keypair is deterministically generated from `entropy`.
+// If the generated keypair is invalid, a new seed will be generated by hashing `entropy` to try again.
+func deriveKeyPair(entropy []byte) (*PublicKey, *PrivateKey) {
+	const (
+		irrPolys  = sysN/8 + (1<<gfBits)*4
+		seedIndex = sysN/8 + (1<<gfBits)*4 + sysT*2
+		permIndex = sysN / 8
+		sBase     = 32 + 8 + irrBytes + condBytes
+	)
+
+	var (
+		pk [PublicKeySize]byte
+		sk [PrivateKeySize]byte
+	)
+
+	seed := [33]byte{64}
+	r := [sysN/8 + (1<<gfBits)*4 + sysT*2 + 32]byte{}
+
+	f := [sysT]gf{}
+	irr := [sysT]gf{}
+	perm := [1 << gfBits]uint32{}
+	pi := [1 << gfBits]int16{}
+	pivots := uint64(0xFFFFFFFF)
+
+	copy(seed[1:], entropy[:])
+
+	for {
+		// expanding and updating the seed
+		err := shake256(r[:], seed[0:33])
+		if err != nil {
+			panic(err)
```

Can we return the error in `DeriveKeypair()`.

I would say `panic()` these code paths should be discouraged.